TH Tracker Update

tehkrizz opened the pull request:
TH Tracker Battlemod compatibility.

tehkrizz committed:
Merge pull request #3 from Windower/dev

Refreshing fork

tehkrizz committed:
TH Tracker update

Updated TH Tracker to new API and made it compatible with Battlemod

tehkrizz committed:
Actual Version Modified

Updated the actual newest version to be Battlemod compatible.

tehkrizz committed:
Version

Forgot to update the version number in this commit

cairthenn commented on addons/thtracker/thtracker.lua:
```diff
@@ -24,162 +24,83 @@
+windower.register_event('incoming text', function(original, new, color)
+	original = original:strip_format()
+	if string.find(original, "Treasure Hunter") and not string.find(original, "TH Tracker:") then
+		name, count = original:match('Additional effect: Treasure Hunter effectiveness against the (.*) increases to (%d+)')
```
I'd recommend "local name, count" so your variables don't exist in the global namespace. Name and count are very generic, so (future) issues could easily arise.

tehkrizz replied:
I'm actually in the middle of correcting the issue with the mob being killed at least having the same name. I want to eventually do better matching to support multiple targets, but that is beyond me right now. I should have the scopes corrected shortly.

cairthenn commented on addons/thtracker/thtracker.lua:
```diff
@@ -24,162 +24,83 @@
+		count = nil
+    end
+
+    deadmob = original:match('%w+ defeats the (.*).')
```
Just another local variable change for you..

`local deadmob = original:match('%w+ defeats the (.*).') or original:match('%w+ defeats (.*).')`

cairthenn commented on addons/thtracker/thtracker.lua:
```diff
@@ -24,162 +24,83 @@
+        th:show()
+		mob = nil
+		deadmob = nil
+		count = nil
```
You can probably delete the following entirely:

```
mob = name

mob = nil
deadmob = nil
count = nil
```

cairthenn reviewed (commented):
There are just a few minor changes regarding variable scope I think you should make before this gets merged.

I think the core functionality of the addon could use an overhaul since the text matching makes no guarantees which mob is having their TH upgraded or being killed, but since you aren't changing that from how it was, it's outside the scope of this review.

tehkrizz committed:
Dead Mob Name Check

Prevents it from removing the entry if a different named enemy dies.
Also added a reset and hide on zone.

z16 commented on addons/thtracker/data/settings.xml:
```diff
@@ -1,9 +1,50 @@
+            </stroke>
+        </text>
+    </global>
 </settings>
```
You should remove the settings file entirely. See my comments on the outdated PR.

tehkrizz commented:
Just commited another change.

Line 54 gives it text on init so the 'th show' command has something to display after loading.

The message checks are now in different statements with returns because they were conflicting with each other at one point. I tried to make the variables local (mob is really the only one that needs to be global), but it completely stopped working. I'm open to suggestions.

z16 commented on addons/thtracker/thtracker.lua:
```diff
@@ -61,33 +62,51 @@ windower.register_event('addon command', function(command, ...)
-        th:hide()
-    end
+	original = original:strip_format()
+	if string.find(original, "Treasure Hunter") then
```
The `string.find` here is not saving much in terms of efficiency, compared to the latter match. And since the latter match is much more precise, I'd remove this if check entirely, along with the `return` at the end of this inner scope.

z16 commented on addons/thtracker/thtracker.lua:
```diff
@@ -61,33 +62,51 @@ windower.register_event('addon command', function(command, ...)
+		return
+	end
+	
+	if string.find(original, "defeats") then
```
Same as above, this check can be removed completely

z16 commented on addons/thtracker/thtracker.lua:
```diff
@@ -61,33 +62,51 @@ windower.register_event('addon command', function(command, ...)
+		name, count = original:match('Additional effect: Treasure Hunter effectiveness against the (.*) increases to (%d+).')
+		if not name and not count then
+			name, count = original:match('Additional effect: Treasure Hunter effectiveness against (.*) increases to (%d+).')
+		end
```
While it would be more expensive, you could consider using a regex match for this. It would save you the double match and if check, so you'd effectively reduce this code (as well as the lower part) to one line.

tehkrizz replied:
That was my original goal, but I have issues with Lua regex. I'll investigate further.

tehkrizz replied:
I've managed to get it to find it with or without the string, but it adds 'the ' before the mob's name in the variable. Not the desired result. I'm currently using:
`against[%s%a%a%a]- (.*) increases`

I can't seem to wrap my head around how to ignore the 'the'.

z16 replied:
You can't do that with Lua patterns, sadly, so either you keep it the way it's in this PR, or you change to regex:
```lua
name, count = select(2, unpack(windower.regex.match(original, 'Additional effect: Treasure Hunter effectiveness against (?:the)? (.*) increases to (\\d+).')[1]))
```

The syntax is a bit more verbose, since `windower.regex.match` returns a different structure for all matches and packs it into a table, so it looks like this.

tehkrizz replied:
The windower.regex was exactly what I was looking for (obviously didn't look hard enough). That will be so much better as it should be similar, if not the same, to what I use for JavaScript. Thanks.

z16 commented on addons/thtracker/thtracker.lua:
```diff
@@ -27,11 +27,12 @@
 
```
You should update the copyright notice for the current year, as well as replace the `(c)` with `©`.

z16 reviewed (approved):
Added some suggestions as well. All rather minor, overall this looks good to me.

tehkrizz committed:
Requested changes

Copyright updated. Variables local where possible. Removed find checks
around matches.
Updated match pattern to maintain BattleMod compatibility, but the
current iteration requires a gsub to get rid of 'the ' if it is present.

tehkrizz committed:
Deleted settings.xml

Forgot to remove this from respository.
